### PR TITLE
process nested obligations in autoderef

### LINKED ISSUE #53843: Method resolution fails to find inherent method on custom Deref type
This cannot find the `poll` method on `Pin`:

```rust
use std::ops::Deref;

pub struct Pin<P>(P);

impl<P, T> Deref for Pin<P> where
    P: Deref<Target = T>,
{
    type Target = T;
    fn deref(&self) -> &T {
        &*self.0
    }
}

impl<'a, F> Pin<&'a mut F> {
    fn poll(self) {}
}

fn test(pin: Pin<&mut ()>) {
    pin.poll()
}
```

https://play.rust-lang.org/?gist=6e3b3f6f3aa4e8a49cb45480f4dd7e7a&version=stable&mode=debug&edition=2015

This clearly should compile.

This bug may be a blocker on changing the Pin API to a composeable form.

cc @nikomatsakis @eddyb @arielb1 @cramertj

### PR DESCRIPTION
**Pull request by arielb1**
Fixes #53843.

r? @nikomatsakis

### CODE REVIEW

**Comment by eddyb** [comment]
IIUC, a fix isn't needed because the original code can be switched to use `P::Target` instead of a type parameter `T` used with `P: Deref<Target = T>`.

**Comment by arielb1** [comment]
@eddyb 

I agree that this makes the bug workaroundable and therefore less urgent. However, this is still a bug that should be fixed..

**Review by nikomatsakis — COMMENTED** [review]
At first read, seems reasonable (for what it is). Should we check perf? It feels like these are some common operations.

**Comment by arielb1** [comment]
@nikomatsakis 

Sure, a perf run would be nice.

**Comment by arielb1** [comment]
Try build failure looks quite spurious to my eyes

**Comment by nikomatsakis** [comment]
Yeah, the try build failure does look spurious. I'm not sure how to rerun.

@bors try-

**Comment by matthiaskrgr** [comment]
Are all these submodule changes intended?

**Comment by nikomatsakis** [comment]
@rust-timer build 0b9d15f8dc84ec750e5ac0c3385cdd38c0d6cf70

**Comment by nikomatsakis** [comment]
Perf results look pretty neutral.

**Comment by TimNN** [comment]
Ping form triage @nikomatsakis: Since "Perf results look pretty neutral" can this PR move forward?

**Review by nikomatsakis — COMMENTED** [review]
Argh sorry @arielb1 I let this slip through! I have this old review with a few questions...I'll just post it.

File: src/librustc_typeck/check/method/probe.rs [hunk 1]
```diff
@@ -295,21 +339,63 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+    })
+}
 
-        let mut autoderef = self.autoderef(span, self_ty).include_raw_pointers();
+fn create_steps_inner<'a, 'gcx, 'tcx>(tcx: TyCtxt<'a, 'gcx, 'gcx>,
+                                      span: Span,
+                                      pe_and_self_ty: Canonical<'gcx, (ParamEnv<'gcx>, Ty<'gcx>)>)
+                                      -> CreateStepsResult<'gcx>
+{
+    tcx.infer_ctxt().enter(|ref infcx| {
```
[the remark this answers is not in the document]
Comment (arielb1) [reply src/librustc_typeck/check/method/probe.rs]: Sure enough. Just me being lazy.

File: src/librustc_typeck/check/method/probe.rs [hunk 2]
```diff
@@ -247,42 +254,79 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
-                     unsize: false,
-                 }]
+            // FIXME: don't make an inference context for nothing here
+            self.tcx.global_tcx().infer_ctxt().enter(|ref infcx| {
+                let ((_, self_ty), canonical_inference_vars) =
+                    infcx.instantiate_canonical_with_fresh_inference_vars(
+                        span, &param_env_and_self_ty);
+                CreateStepsResult {
+                    steps: vec![CandidateStep {
+                        self_ty: do_make_query_result(infcx, &canonical_inference_vars, self_ty),
```
[the remark this answers is not in the document]
Comment (arielb1) [reply src/librustc_typeck/check/method/probe.rs]: This is just me being lazy.

**Comment by nikomatsakis** [comment]
@arielb1 do you want to try and cleanup those little bits before we land, or would you rather land as is? If the latter, mind opening a FIXME issue (and adding corresponding comments into the code)?

**Comment by nikomatsakis** [comment]
r=me either way

**Comment by arielb1** [comment]
I'll fix them and then land. I'll try to do this soon.

**Comment by nikomatsakis** [comment]
@arielb1 sounds good!

**Comment by TimNN** [comment]
Ping from triage @arielb1: What is the status of this PR?

**Comment by arielb1** [comment]
@TimNN 

I'll do it when I have a bit more time.

**Comment by TimNN** [comment]
Ping from triage! Thanks for your PR, @arielb1. It's been a while since this PR was last updated, so I'm closing it as inactive for now, following our [guidelines](https://forge.rust-lang.org/triage-procedure.html). Thanks for your contributions and please re-open when / if you get a chance to work on this again!

File: src/librustc_typeck/check/method/probe.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (arielb1) [reply src/librustc_typeck/check/method/probe.rs]: No, we also want a QueryResponse while we have a Canonical Query, so we need to use some function.

**Comment by arielb1** [comment]
addressed comments

**Comment by arielb1** [comment]
github is not letting me reopen this PR because "the deref-query branch was force-pushed or recreated"

**Comment by eddyb** [comment]
@arielb1 The way to fix that is to force push the original commit (if you can find its hash anywhere), reopen the PR, then force push the new state of the branch.

**Comment by nikomatsakis** [comment]
@arielb1 nice, will look asap =)

**Review by nikomatsakis — COMMENTED** [review]
Sorry for the delay. Been a crazy week. Basically r=me, but I'm wondering whether it makes sense to do a crater run for something this potentially far-reaching. What do you think, @arielb1 ?

**Comment by Mark-Simulacrum** [comment]
Not sure what you mean by mid-release crater runs, but I think it may make sense to run crater here. I'll get the ball rolling, if we decide to go ahead, @bors try

**Comment by nikomatsakis** [comment]
@arielb1 I took the liberty of rebasing this for you

**Comment by arielb1** [comment]
@nikomatsakis

I already had a rebased commit

**Comment by nikomatsakis** [comment]
@craterbot run start=master#96d1334e567237b1507cd277938e7ae2de75ff51 end=try#0b8c5828f68d7b2840b81fa51239346dc97fef5a mode=check-only

**Comment by arielb1** [comment]
Yay!

**Pull request merged**